ortools/linear_solver/xpress_interface.cc

```diff
@@ -1931,9 +1999,104 @@ void XpressInterface::AddSolutionHintToOptimizer() {
     val[i] = solver_->solution_hint_[i].second;
   }
-  if (int status = XPRSaddmipsol(mLp, len, val.get(), colind.get(), "USER_HINT")) {
-    LOG(WARNING) << "Failed to set solution hint.";
+  XPRSaddhint(mLp, len, val.get(), colind.get(), "USER_HINT");
+}
+
+void XpressInterface::SetCallback(MPCallback* mp_callback) {
+  if (callback_ != nullptr) {
+    // replace existing callback by removing it first
+    CHECK_STATUS(XPRSremovecboptnode(mLp, XpressOptNodeCallbackImpl, NULL));
+  }
+  callback_ = mp_callback;
+}
+
+// NOTE(user): This function must have this exact API, because we are passing
+// it to XPRESS as a callback.
+void XPRS_CC XpressOptNodeCallbackImpl(XPRSprob cbprob, void* cbdata, int* p_infeasible) {
+  MPCallbackWithXpressContext* const callback_with_context = static_cast<MPCallbackWithXpressContext*>(cbdata);
+  CHECK(callback_with_context != nullptr);
+  CHECK(callback_with_context->context != nullptr);
+  CHECK(callback_with_context->callback != nullptr);
+  if (callback_with_context->context->UpdateFromXpressState(cbprob)) {
+    // Only run user's callback if the context has changed
+    callback_with_context->callback->RunCallback(
+        callback_with_context->context);
   }
 }
 
+XpressMPCallbackContext::XpressMPCallbackContext(XPRSprob& mLp, bool mip,
+                                                 bool maximize, int num_vars)
+    : mLp_(mLp),
+      mip_(mip),
+      maximize_(maximize),
+      num_vars_(num_vars),
+      variable_values_(num_vars_, XPRS_NAN),
+      objective_value_(XPRS_NAN),
+      best_objective_bound_(XPRS_NAN),
+      num_nodes_(0),
+      num_user_sols_(0) {}
+
+MPCallbackEvent XpressMPCallbackContext::Event() {
+  return MPCallbackEvent::kMipSolution; // this is the only supported callback for now
+}
+
+bool XpressMPCallbackContext::CanQueryVariableValues() {
+  return !std::isnan(objective_value_);
+}
+
+double XpressMPCallbackContext::VariableValue(const MPVariable* variable) {
+  return variable_values_[variable->index()];
+}
+
+double XpressMPCallbackContext::SuggestSolution(
+    const absl::flat_hash_map<const MPVariable*, double>& solution) {
+  // Currently the XPRESS API does not handle clearing out previous hints
+  const std::size_t len = solution.size();
+  if (len == 0) {
+    // hint is empty, do nothing
+    return XPRS_NAN;
+  }
+  unique_ptr<int[]> colind(new int[len]);
+  unique_ptr<double[]> val(new double[len]);
+  int i = 0;
+  for (const auto& [var, value] : solution) {
+    colind[i] = var->index();
+    val[i] = value;
+    ++i;
+  }
+  // use a new name so new solutions don't override old ones
+  XPRSaddhint(mLp_, len, val.get(), colind.get(),
+              "USER_CB_HINT_" + std::to_string(++num_user_sols_));
+
+  // XPRESS doesn't guarantee when it will test the suggested solution
+  // So we return the last known objective value but with no guarantee that it
+  // is computed on the suggested solution
+  return objective_value_;
```

Comment:
Does this make sense ?

Reply:
from [here](https://www.fico.com/fico-xpress-optimization/docs/latest/solver/optimizer/HTML/XPRSaddmipsol.html): 
"The function returns immediately after passing the solution to the Optimizer. The solution is placed in a pool until the Optimizer is able to analyze the solution during a MIP solve."

would you rather we return NaN instead ?

Reply:
There is no guarantee Xpress accepts the solution, even if it is feasible. So both NaN and the best objective seen so far are not a very good match. Maybe NaN is the best to indicate that we don't know.
In any case, the value returned should be the same for an empty and a non-empty solution.

Reply:
I will use NaN